core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala

```diff
@@ -0,0 +1,241 @@
+      appenders.nextElement().asInstanceOf[Appender].getLayout()
+    } else {
+      new PatternLayout(conf.get(DRIVER_LOG_LAYOUT))
+    }
+    val fa = new Log4jFileAppender(layout, localLogFile)
+    fa.setName(DriverLogger.APPENDER_NAME)
+    LogManager.getRootLogger().addAppender(fa)
+    logInfo(s"Added a local log appender at: ${localLogFile}")
+  }
+
+  def startSync(hadoopConf: Configuration): Unit = {
+    try {
+      val appId = Utils.sanitizeDirName(conf.getAppId)
+      hdfsLogFile = {
+        val rootDir = conf.get(DRIVER_LOG_DFS_DIR.key,
+          "/tmp/driver_logs").split(",").head
+        FileUtils.getFile(rootDir, appId, DRIVER_LOG_FILE).getAbsolutePath()
+      }
+      hadoopConfiguration = hadoopConf
+      fileSystem = FileSystem.get(hadoopConf)
+
+      // Setup a writer which moves the local file to hdfs continuously
+      if (syncToDfs) {
+        writer = Some(new HdfsAsyncWriter())
+        logInfo(s"The local driver log file is being synced to spark dfs at: ${hdfsLogFile}")
+      }
+    } catch {
+      case e: Exception =>
+        logError(s"Could not sync driver logs to spark dfs", e)
+    }
+  }
+
+  def stop(): Unit = {
+    try {
+      writer.map(_.closeWriter())
+      if (syncToYarn) {
+        moveToYarnAppDir()
+      }
+    } catch {
+      case e: Exception =>
+        logError(s"Error in persisting driver logs", e)
+    } finally {
+      try {
+        JavaUtils.deleteRecursively(FileUtils.getFile(localLogFile).getParentFile())
+      } catch {
+        case e: Exception =>
+          logError(s"Error in deleting local driver log dir", e)
+      }
+    }
+  }
+
+  private def moveToYarnAppDir(): Unit = {
+    try {
+      val appId = Utils.sanitizeDirName(conf.getAppId)
+      val yarnConf = new YarnConfiguration(hadoopConfiguration)
+      val rootDir = yarnConf.get(YarnConfiguration.NM_REMOTE_APP_LOG_DIR)
+      if (rootDir != null && !rootDir.isEmpty()) {
+        var parentDir = if (UserGroupInformation.getCurrentUser() != null) {
+          FileUtils.getFile(
+            rootDir,
+            UserGroupInformation.getCurrentUser().getShortUserName(),
+            "logs",
+            appId)
+        } else {
+          FileUtils.getFile(rootDir, "logs", appId)
+        }
+        if (syncToDfs) {
+          fileSystem.rename(new Path(hdfsLogFile), new Path(parentDir.getAbsolutePath()))
+          fileSystem.delete(new Path(FileUtils.getFile(hdfsLogFile).getParent()), true)
+          logInfo(
+            s"Moved the driver log file to: ${parentDir.getAbsolutePath()}")
+        } else {
+          fileSystem.copyFromLocalFile(true,
+            new Path(localLogFile),
+            new Path(parentDir.getAbsolutePath()))
+          logInfo(
+            s"Moved the local driver log file to spark dfs at: ${parentDir.getAbsolutePath()}")
+        }
+      }
+    } catch {
+      case nse: NoSuchElementException => logWarning("Couldn't move driver log" +
```

Comment:
Move code after `=>` to next line. Where is this exception thrown btw? The log message is very specific...

Reply:
This exception can be thrown in conf.getAppId. The scenario is if there is an exception before Spark application starts, then this exception will be thrown while trying to close the logger.